File: backend/internal/models/quota.go
package models

import (
	"time"
	"gorm.io/gorm"
)

// UserQuota 用户配额
type UserQuota struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	UserID      uint   `gorm:"not null" json:"userId"`
	Path        string `gorm:"not null" json:"path"`
	SoftLimit   uint64 `json:"softLimit"`   // 软限制（字节）
	HardLimit   uint64 `json:"hardLimit"`   // 硬限制（字节）
	UsedSpace   uint64 `json:"usedSpace"`   // 已用空间（字节）
	GracePeriod int    `json:"gracePeriod"` // 宽限期（天）
	FilesUsed   uint64 `json:"filesUsed"`   // 已用文件数
	FilesSoft   uint64 `json:"filesSoft"`   // 文件数软限制
	FilesHard   uint64 `json:"filesHard"`   // 文件数硬限制

	// 关联用户
	User *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

// GroupQuota 组配额
type GroupQuota struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	GroupID     uint   `gorm:"not null" json:"groupId"`
	Path        string `gorm:"not null" json:"path"`
	SoftLimit   uint64 `json:"softLimit"`   // 软限制（字节）
	HardLimit   uint64 `json:"hardLimit"`   // 硬限制（字节）
	UsedSpace   uint64 `json:"usedSpace"`   // 已用空间（字节）
	GracePeriod int    `json:"gracePeriod"` // 宽限期（天）
	FilesUsed   uint64 `json:"filesUsed"`   // 已用文件数
	FilesSoft   uint64 `json:"filesSoft"`   // 文件数软限制
	FilesHard   uint64 `json:"filesHard"`   // 文件数硬限制
}

// QuotaReport 配额报告
type QuotaReport struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name       string    `json:"name"`       // 用户名或组名
	Type       string    `json:"type"`       // user, group
	Path       string    `json:"path"`
	UsedSpace  uint64    `json:"usedSpace"`
	SoftLimit  uint64    `json:"softLimit"`
	HardLimit  uint64    `json:"hardLimit"`
	UsedPercent float64  `json:"usedPercent"` // 使用率百分比
	Status     string    `json:"status"`     // ok, warning, exceeded, grace
	FilesUsed  uint64    `json:"filesUsed"`
	FilesSoft  uint64    `json:"filesSoft"`
	FilesHard  uint64    `json:"filesHard"`
	GeneratedAt time.Time `json:"generatedAt"`
}

// QuotaAlert 配额告警
type QuotaAlert struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	UserID     uint   `gorm:"not null" json:"userId"`
	GroupID    uint   `json:"groupId,omitempty"`
	Type       string `json:"type"`       // user, group
	Path       string `json:"path"`
	AlertType  string `json:"alertType"`  // soft_limit, hard_limit, grace_period
	Severity   string `json:"severity"`   // warning, critical
	Message    string `json:"message"`
	Resolved   bool   `json:"resolved"`
	ResolvedAt *time.Time `json:"resolvedAt,omitempty"`
}

File: frontend/src/types/quota.ts
// Quota Management Type Definitions

export interface UserQuota {
  id: number
  userId: number
  user: User
  path: string
  softLimit: number
  hardLimit: number
  usedSpace: number
  filesUsed: number
  filesSoft: number
  filesHard: number
  gracePeriod: number
  createdAt: string
  updatedAt: string
}

export interface GroupQuota {
  id: number
  groupId: number
  group: Group
  path: string
  softLimit: number
  hardLimit: number
  usedSpace: number
  filesUsed: number
  filesSoft: number
  filesHard: number
  gracePeriod: number
  createdAt: string
  updatedAt: string
}

export interface QuotaReport {
  name: string
  type: 'user' | 'group'
  path: string
  usedSpace: number
  softLimit: number
  hardLimit: number
  usedPercent: number
  status: 'ok' | 'warning' | 'exceeded'
  filesUsed: number
  filesSoft: number
  filesHard: number
  generatedAt: string
}

export interface QuotaAlert {
  id: number
  userId: number | null
  groupId: number | null
  type: 'user' | 'group' | 'user_quota' | 'disk_space'
  path: string
  alertType: 'soft_limit' | 'hard_limit'
  severity: 'warning' | 'critical'
  message: string
  resolved: boolean
  createdAt: string
  updatedAt: string
  // UI 扩展字段
  name?: string
  enabled?: boolean
  threshold?: number
  duration?: number
  condition?: string
  actions?: string
}

export interface User {
  id: number
  username: string
  displayName: string
  email: string
  role: string
}

export interface Group {
  id: number
  name: string
  description: string
}

export interface QuotaSettings {
  path: string
  softLimit: number
  hardLimit: number
  filesSoft: number
  filesHard: number
  gracePeriod: number
  enabled: boolean
}

export const QUOTA_STATUS_COLORS = {
  ok: 'green',
  warning: 'orange',
  exceeded: 'red'
}

export const QUOTA_TYPE_LABELS = {
  user: '用户配额',
  group: '组配额'
}

export const ALERT_TYPE_LABELS = {
  soft_limit: '软限制',
  hard_limit: '硬限制'
}

export const SEVERITY_COLORS = {
  warning: 'orange',
  critical: 'red'
}

// Helper functions for quota calculations
export function calculateQuotaPercent(used: number, limit: number): number {
  if (limit === 0) return 0
  return (used / limit) * 100
}

export function getQuotaStatus(used: number, softLimit: number, hardLimit: number): 'ok' | 'warning' | 'exceeded' {
  if (hardLimit > 0 && used > hardLimit) return 'exceeded'
  if (softLimit > 0 && used > softLimit) return 'warning'
  return 'ok'
}

export function formatBytes(bytes: number): string {
  if (bytes === 0) return '0 B'
  const k = 1024
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB']
  const i = Math.floor(Math.log(bytes) / Math.log(k))
  return `${(bytes / Math.pow(k, i)).toFixed(1)} ${sizes[i]}`
}

export function formatFilesCount(count: number): string {
  return count.toLocaleString()
}